### ISSUE TITLE
Try replacing more call sites of `AnyProvider` with `DataProvider<M>` and then remove `with_any_provider` ctors

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#3947]
I think some of the tutorials can be written using `DataProvider<M: KeyedDataMarker>` instead of `AnyProvider`. However, in order to act on any particular data request (such as overwriting it), without specialization I think they'll still need to translate via `AnyPayload`. However, I think we can maybe get away in most cases with only `AnyPayload` without full `AnyProvider`.

**Comment by robertbastian** [comment]
What value does `AnyPayload` give us compared to `DataPayload<AnyMarker>`? `AnyPayload` and `AnyResponse` are types that don't exist on the buffer side, so if we want to phase out `AnyProvider`, do we still need them? Could the Yokeable for `AnyMarker` be something like `Rc<dyn Any>` instead of a nested payload?

**Comment by sffc** [comment]
There's only one reason: `AnyPayload` can be created from a `&'static dyn Any` whereas `DataPayload<AnyMarker>` always allocates into a `Box<dyn Any>`.

Not sure if I ever attempted making `AnyMarker::Yokeable` be an enum between `&'static dyn Any` and `Box<dyn Any>` but if that works then maybe we can phase out `AnyPayload`.

**Comment by robertbastian** [comment]
> whereas `DataPayload<AnyMarker>` always allocates into a Box<dyn Any>.

`DataPayload<AnyMarker>` uses `AnyPayload` as its `Yokeable`, so it is allocation-free. The only difference compared to using `AnyPayload` directly is that it gets wrapped in an owned `DataPayload`.

**Comment by robertbastian** [comment]
When we remove `AnyPayload` from the public API we can change it to use `Box` instead of `SelectedRc` internally, as it doesn't need to be `Clone` when used as a `Yokeable`. This could give some performance improvement.

**Comment by sffc** [comment]
Related: #1893, #3935

**Comment by robertbastian** [comment]
It looks like most uses of `AnyProvider` are `AnyPayloadProvider` or custom providers that return fixed payloads for multiple keys. We should be able to replace those by `DataProvider`.

We can unify `AnyProvider` and `DynamicDataProvider<AnyMarker>` if we can get `AnyMarker`'s `Yokeable` to be `dyn Any`. This would remove the extra wrapping and reference counting, and would allow downcasting without cloning.

**Comment by sffc** [comment]
Action is for @sffc or @Manishearth to try migrating `AnyPayload` to `DataPayload<AnyMarker>` which will enable the cleanup.

**Comment by Manishearth** [comment]
Working on part 1

**Comment by sffc** [comment]
Did we discuss whether we can get rid of `try_new_with_any_provider` constructors?

**Comment by Manishearth** [comment]
I think we wanted to? I don't remember.

**Comment by robertbastian** [comment]
> I'm actually not clear on what this means, AnyMarker's yokeable is AnyPayload.

It would be nice if `AnyMarker`'s yokeable would be `dyn Any`

**Comment by jedel1043** [comment]
> Did we discuss whether we can get rid of try_new_with_any_provider constructors?

One of the nicer things about `*_with_any_provider` was that any bound changes in the `*_unstable` constructor could be "absorbed" by the `any` constructor, making it really easy to update the icu4x version using the same baked data, and doing the data regeneration in a later point in time. Would there be an alternative to this with the removal of `AnyProvider`? Or was it deemed not useful for some other reasons I'm not seeing right now?

**Comment by sffc** [comment]
That was the theory, but one cannot simply use baked data across ICU4X minor versions, because data structs are subject to change, and one cannot build the structs against the old version and plug into the new version because the TypeIds are not going to match. So, the whole concept I think was flawed to begin with. 

We still have buffer provider for clients who want the stability, and that works fine: it makes use of the postcard format which we explicitly maintain for this purpose.

**Comment by Manishearth** [comment]
Yeah, the anyprovider design predates baked, and we primarily figured it would be useful for manually constructed data, but that's not really a major use case so far.
